File: src/main/java/self/ed/main/RunSolver.java
package self.ed.main;

import static java.lang.System.currentTimeMillis;
import static java.util.Arrays.stream;

import static self.ed.main.DataDirs.READY_DIR;
import static self.ed.util.FileUtils.readFile;
import static self.ed.util.FileUtils.streamFiles;
import static self.ed.util.SudokuUtils.fromString1D;
import static self.ed.util.SudokuUtils.toString2D;

import java.io.File;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.atomic.AtomicLong;
import java.util.stream.Collectors;
import java.util.stream.Stream;

import org.apache.commons.lang3.StringUtils;
import org.apache.commons.lang3.tuple.Pair;

import self.ed.solver.CleverSolver;
import self.ed.util.SudokuUtils;

public class RunSolver {

  public static void main(String[] args) {
    solveSingleComplexSudoku();
//    solveAllReadySudoku();
  }

  private static void solveSingleComplexSudoku() {
    String sudokuStr = ".......9......8.2.7.3.54.....52.................6....88....3..7.9....6..6...8...4";
    int[][] sudoku = fromString1D(sudokuStr);
    long start = currentTimeMillis();
    int[][] solution = new CleverSolver(sudoku).solve();
    System.out.println(toString2D(solution));
    System.out.println("Time: " + (currentTimeMillis() - start));
  }

  private static void solveAllReadySudoku() {
    List<File> files = streamFiles(READY_DIR).sorted().toList();
    Map<String, Long> counts = files.stream()
        .collect(Collectors.toMap(File::getName, RunSolver::countLines));
    long count = counts.values().stream().mapToLong(i -> i).sum();
    System.out.println(counts);
    System.out.println("Total count: " + count); // 1006520
    AtomicLong counter = new AtomicLong();
    long start = currentTimeMillis();
    Map<String, Pair<Long, Long>> statistics = new LinkedHashMap<>();
    for (File file : files) {
      long fileStart = currentTimeMillis();
      long fileCount = streamLines(file).map(SudokuUtils::fromString1D).peek(sudoku -> {
        long progress = counter.incrementAndGet();
        if (progress % 1_000 == 0) {
          System.out.println("progress: " + progress + " / " + ((currentTimeMillis() - start) / 1000) + "s");
        }
        if (progress > 10_000) {
          throw new RuntimeException("stopped");
        }
        new CleverSolver(sudoku).solve();
      }).count();
      long fileEnd = currentTimeMillis();
      statistics.put(file.getName(), Pair.of(fileCount, fileEnd - fileStart));
    }
    long end = currentTimeMillis();
    System.out.println("Total count: " + count); // 1006520
    System.out.println("Total time: " + (end - start) / 1000);
    statistics.forEach((file, stat) -> {
      System.out.println(file + ": " + stat.getRight() + " / " + ((double) stat.getRight() / stat.getLeft()));
    });
  }

  /*
Clever:
progress: 1000 / 4s
progress: 2000 / 8s
progress: 3000 / 12s
progress: 4000 / 16s
progress: 5000 / 19s
progress: 6000 / 22s
progress: 7000 / 26s
progress: 8000 / 29s
progress: 9000 / 33s
progress: 10000 / 36s
Simple:
progress: 1000 / 41s
progress: 2000 / 87s
progress: 3000 / 114s
progress: 4000 / 137s
progress: 5000 / 159s
progress: 6000 / 179s
progress: 7000 / 202s
progress: 8000 / 227s
progress: 9000 / 250s
progress: 10000 / 274s
   */

  private static Long countLines(File file) {
    return streamLines(file).count();
  }

  private static Stream<String> streamLines(File file) {
    return stream(readFile(file).split("\n"))
        .filter(StringUtils::isNotEmpty);
  }

}

File: cpp-sudoku/src/main/RunSolver.cpp
#include <iostream>
#include <chrono>
#include <map>
#include <algorithm>
#include <sstream>
#include "../solver/Solver.hpp"
#include "../util/SudokuUtils.hpp"
#include "../util/FileUtils.hpp"
#include "DataDirs.hpp"

using namespace just::demo::solver;
using namespace just::demo::util;
using namespace just::demo::main;

void solveSingleComplexSudoku() {
    std::string sudokuStr = ".......9......8.2.7.3.54.....52.................6....88....3..7.9....6..6...8...4";
    auto sudoku = SudokuUtils::fromString1D(sudokuStr);
    
    auto start = std::chrono::steady_clock::now();
    auto solution = Solver(sudoku).solve();
    auto end = std::chrono::steady_clock::now();
    
    std::cout << SudokuUtils::toString2D(solution) << std::endl;
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::cout << "Time: " << duration.count() << "ms" << std::endl;
}

long countLines(const std::filesystem::path& file) {
    std::string content = FileUtils::readFile(file);
    return std::count(content.begin(), content.end(), '\n') + 1;
}

void solveAllReadySudoku() {
    auto files = FileUtils::streamFiles(DataDirs::READY_DIR);
    std::sort(files.begin(), files.end());
    
    std::map<std::string, long> counts;
    for (const auto& file : files) {
        counts[file.filename().string()] = countLines(file);
    }
    
    long totalCount = 0;
    for (const auto& [_, count] : counts) {
        totalCount += count;
    }
    
    std::cout << "Total count: " << totalCount << std::endl;
    
    long counter = 0;
    auto start = std::chrono::steady_clock::now();
    
    for (const auto& file : files) {
        auto fileStart = std::chrono::steady_clock::now();
        std::string content = FileUtils::readFile(file);
        
        std::istringstream iss(content);
        std::string line;
        long fileCount = 0;
        
        while (std::getline(iss, line)) {
            if (!line.empty()) {
                ++counter;
                if (counter % 1000 == 0) {
                    auto now = std::chrono::steady_clock::now();
                    auto elapsed = std::chrono::duration_cast<std::chrono::seconds>(now - start);
                    std::cout << "progress: " << counter << " / " << elapsed.count() << "s" << std::endl;
                }
                if (counter > 10000) {
                    throw std::runtime_error("stopped");
                }
                
                auto sudoku = SudokuUtils::fromString1D(line);
                Solver(sudoku).solve();
                ++fileCount;
            }
        }
        
        auto fileEnd = std::chrono::steady_clock::now();
        auto fileDuration = std::chrono::duration_cast<std::chrono::milliseconds>(fileEnd - fileStart);
        std::cout << file.filename().string() << ": " << fileDuration.count() << "ms / " 
                 << (static_cast<double>(fileDuration.count()) / fileCount) << "ms per puzzle" << std::endl;
    }
    
    auto end = std::chrono::steady_clock::now();
    auto totalDuration = std::chrono::duration_cast<std::chrono::seconds>(end - start);
    std::cout << "Total time: " << totalDuration.count() << "s" << std::endl;
}

int main() {
    solveSingleComplexSudoku();
    // solveAllReadySudoku(); // Uncomment to run all puzzles
    return 0;
}
